File: packages/node/src/embed.ts
import { ProviderName, EmbeddingOptions, EmbeddingResponse, FallbackItem } from './types.js';
import { PROVIDERS, callEmbeddingAPI } from './providers.js';

export interface OmniEmbedConfig {
  apiKeys?: Partial<Record<ProviderName, string>>;
  defaultFallbackOrder?: ProviderName[];
}

const DEFAULT_EMBEDDING_PRIORITY: ProviderName[] = [
  'huggingface',
  'openai',
  'gemini',
  'cohere',
  'jina',
];

export class OmniEmbed {
  private apiKeys: Partial<Record<ProviderName, string>> = {};
  private defaultFallbackOrder: ProviderName[];

  constructor(config?: OmniEmbedConfig) {
    // Initialize API keys from config or process.env
    for (const provider of DEFAULT_EMBEDDING_PRIORITY) {
      const configKey = PROVIDERS[provider].envKey;
      this.apiKeys[provider] =
        config?.apiKeys?.[provider] ||
        (typeof process !== 'undefined' ? process.env[configKey] : undefined);
    }

    this.defaultFallbackOrder = config?.defaultFallbackOrder || DEFAULT_EMBEDDING_PRIORITY;
  }

  /**
   * Generates a vector embedding using the configured providers and fallback logic.
   * @param prompt The text prompt to embed.
   * @param options Execution options including custom fallbacks, model choice, etc.
   */
  async embed(
    prompt: string,
    options?: EmbeddingOptions & { provider?: ProviderName }
  ): Promise<EmbeddingResponse> {
    const errors: Array<{ provider: ProviderName; model: string; error: string }> = [];

    // 1. Resolve the sequence of provider/model attempts
    let queue: FallbackItem[] = [];

    if (options?.fallback && options.fallback.length > 0) {
      // Use user-provided fallback sequence
      queue = options.fallback.map((item) => ({
        provider: item.provider,
        model: item.model || PROVIDERS[item.provider].defaultEmbeddingModel,
      }));
    } else {
      // Build fallback queue dynamically
      const primaryProvider = options?.provider;
      const primaryModel = options?.model;

      if (primaryProvider) {
        queue.push({
          provider: primaryProvider,
          model: primaryModel || PROVIDERS[primaryProvider].defaultEmbeddingModel,
        });
      }

      // Find all other providers that have keys available, in priority order
      const availableProviders = this.defaultFallbackOrder.filter((p) => {
        if (p === primaryProvider) return false;
        const key = this.apiKeys[p];
        return !!key && !!PROVIDERS[p].defaultEmbeddingModel;
      });

      for (const p of availableProviders) {
        queue.push({
          provider: p,
          model: PROVIDERS[p].defaultEmbeddingModel,
        });
      }
    }

    // If queue is empty, fallback to default order
    if (queue.length === 0) {
      if (options?.provider) {
        queue.push({
          provider: options.provider,
          model: options.model || PROVIDERS[options.provider].defaultEmbeddingModel,
        });
      } else {
        for (const p of this.defaultFallbackOrder) {
          if (PROVIDERS[p].defaultEmbeddingModel) {
            queue.push({
              provider: p,
              model: PROVIDERS[p].defaultEmbeddingModel,
            });
          }
        }
      }
    }

    // 2. Iterate and attempt calls in sequence
    for (const attempt of queue) {
      const { provider, model } = attempt;
      const resolvedModel = model || PROVIDERS[provider].defaultEmbeddingModel;

      if (!resolvedModel) {
        errors.push({
          provider,
          model: '',
          error: `Provider ${provider} does not support embeddings.`,
        });
        continue;
      }

      // Determine API key for this provider
      const apiKey =
        (provider === options?.provider ? options?.apiKey : undefined) ||
        this.apiKeys[provider];

      if (!apiKey) {
        errors.push({
          provider,
          model: resolvedModel,
          error: `Missing API key for provider ${provider}. Check your env (${PROVIDERS[provider].envKey}) or constructor configuration.`,
        });
        continue;
      }

      try {
        const embedding = await callEmbeddingAPI(provider, resolvedModel, prompt, apiKey);

        return {
          success: true,
          provider,
          model: resolvedModel,
          embedding,
          dimensions: embedding.length,
          errors,
        };
      } catch (err: any) {
        errors.push({
          provider,
          model: resolvedModel,
          error: err?.message || String(err),
        });
      }
    }

    // 3. If all attempts failed
    const fallbackProvider = queue[0]?.provider || 'huggingface';
    return {
      success: false,
      provider: fallbackProvider,
      model: queue[0]?.model || PROVIDERS[fallbackProvider].defaultEmbeddingModel || '',
      embedding: [],
      dimensions: 0,
      errors,
    };
  }
}

File: packages/python/src/omnicall_llm/embed.py
import os
from typing import Dict, Any, List, Optional

from .providers import PROVIDERS, call_embedding_api

DEFAULT_EMBEDDING_PRIORITY = [
    'huggingface',
    'openai',
    'gemini',
    'cohere',
    'jina',
]

class EmbeddingResponse:
    def __init__(
        self,
        success: bool,
        provider: str,
        model: str,
        embedding: List[float],
        dimensions: int,
        errors: Optional[List[Dict[str, str]]] = None
    ):
        self.success = success
        self.provider = provider
        self.model = model
        self.embedding = embedding
        self.dimensions = dimensions
        self.errors = errors if errors is not None else []

    def to_dict(self) -> Dict[str, Any]:
        return {
            "success": self.success,
            "provider": self.provider,
            "model": self.model,
            "embedding": self.embedding,
            "dimensions": self.dimensions,
            "errors": self.errors
        }

    def __repr__(self) -> str:
        return f"EmbeddingResponse(success={self.success}, provider='{self.provider}', model='{self.model}', dimensions={self.dimensions})"

class OmniEmbed:
    def __init__(
        self,
        api_keys: Optional[Dict[str, str]] = None,
        default_fallback_order: Optional[List[str]] = None
    ):
        self.api_keys = {}
        self.default_fallback_order = default_fallback_order or DEFAULT_EMBEDDING_PRIORITY

        for provider in DEFAULT_EMBEDDING_PRIORITY:
            config = PROVIDERS[provider]
            env_key = config['env_key']
            
            # Read from constructor args first, then environment
            self.api_keys[provider] = (
                (api_keys.get(provider) if api_keys else None)
                or os.environ.get(env_key)
            )

    def embed(
        self,
        prompt: str,
        model: Optional[str] = None,
        provider: Optional[str] = None,
        api_key: Optional[str] = None,
        fallback: Optional[List[Dict[str, str]]] = None
    ) -> EmbeddingResponse:
        errors = []
        queue = []

        if fallback:
            # User provided a custom fallback sequence
            for item in fallback:
                p_name = item.get("provider")
                if p_name:
                    queue.append({
                        "provider": p_name,
                        "model": item.get("model") or PROVIDERS[p_name]["default_embedding_model"]
                    })
        else:
            # Build fallback queue dynamically
            if provider:
                queue.append({
                    "provider": provider,
                    "model": model or PROVIDERS[provider]["default_embedding_model"]
                })

            # Append other available providers in priority order
            available_providers = [
                p for p in self.default_fallback_order
                if p != provider and self.api_keys.get(p) and PROVIDERS[p].get("default_embedding_model")
            ]

            for p in available_providers:
                queue.append({
                    "provider": p,
                    "model": PROVIDERS[p]["default_embedding_model"]
                })

        # If queue is empty, fallback to default order
        if not queue:
            if provider:
                queue.append({
                    "provider": provider,
                    "model": model or PROVIDERS[provider]["default_embedding_model"]
                })
            else:
                for p in self.default_fallback_order:
                    if PROVIDERS[p].get("default_embedding_model"):
                        queue.append({
                            "provider": p,
                            "model": PROVIDERS[p]["default_embedding_model"]
                        })

        # Iterate and call in sequence
        for attempt in queue:
            p_name = attempt["provider"]
            m_name = attempt["model"]

            if not m_name:
                errors.append({
                    "provider": p_name,
                    "model": "",
                    "error": f"Provider {p_name} does not support embeddings."
                })
                continue

            # Resolve API Key
            resolved_key = (
                api_key if p_name == provider else None
            ) or self.api_keys.get(p_name)

            if not resolved_key:
                errors.append({
                    "provider": p_name,
                    "model": m_name,
                    "error": f"Missing API key for provider {p_name}. Check your env ({PROVIDERS[p_name]['env_key']}) or constructor configuration."
                })
                continue

            try:
                embedding = call_embedding_api(
                    provider=p_name,
                    model=m_name,
                    prompt=prompt,
                    api_key=resolved_key
                )
                return EmbeddingResponse(
                    success=True,
                    provider=p_name,
                    model=m_name,
                    embedding=embedding,
                    dimensions=len(embedding),
                    errors=errors
                )
            except Exception as e:
                errors.append({
                    "provider": p_name,
                    "model": m_name,
                    "error": str(e)
                })

        # If all failed
        default_provider = queue[0]["provider"] if queue else "huggingface"
        default_model = queue[0]["model"] if queue else PROVIDERS["huggingface"]["default_embedding_model"]
        return EmbeddingResponse(
            success=False,
            provider=default_provider,
            model=default_model,
            embedding=[],
            dimensions=0,
            errors=errors
        )
